### PR TITLE
fix #12634 - port sbt/zinc#979 - add sealedDescendants to zinc sealed children

### LINKED ISSUE #12634: Port zinc changes to sealed child dependencies
Port https://github.com/sbt/zinc/pull/979

## Compiler version

3.0.0 and a956774180d124adc98527863b919c35d5f9db92

## Minimized code

```scala
// A.scala
sealed trait Z
sealed trait A extends Z
class B extends A
class C extends A
// class D extends A
```

```Scala
// App.scala
object App {
  def foo(z: Z) = z match {
    case _: B =>
    case _: C =>
  }
}
```

## Output

if you compile `A` and `App`, uncomment the line in `A` and recompile, then `App` does not recompile

## Expectation

exhaustivity warning `match may not be exhaustive. It would fail on pattern case: _: D`

### PR DESCRIPTION
**Pull request by bishabosha**
Add sealedDescendants method to SymDenotation - get recursive children of a symbol

fixes #12634

### CODE REVIEW

**Review by dwijnand — APPROVED** [review]
LGTM

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala [hunk 1]
```diff
@@ -1614,6 +1614,26 @@ object SymDenotations {
...
+     */
+    final def sealedDescendants(using Context): List[Symbol] =
+      @tailrec
+      def loop(syms: List[Symbol], acc: List[Symbol]): List[Symbol] = syms match
+        case sym :: syms1 =>
+          val notSeen = sym.children.filterConserve(!acc.contains(_))
+          if notSeen.isEmpty then
+            loop(syms1, acc)
+          else
+            loop(syms1 ::: notSeen, acc ::: notSeen)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala:1629]: Appending into a ListBuffer would be more efficient

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/core/SymDenotations.scala]: I have refactored to optimise for the case of a single level deep hierarchy

**Comment by bishabosha** [comment]
added a test for `sealedDescendants` to verify its contents

**Pull request merged**